pcsx2/Frontend/Achievements.cpp

```diff
@@ -0,0 +1,1855 @@
+{
+	static const char* events[] = {"RC_RUNTIME_EVENT_ACHIEVEMENT_ACTIVATED", "RC_RUNTIME_EVENT_ACHIEVEMENT_PAUSED",
+		"RC_RUNTIME_EVENT_ACHIEVEMENT_RESET", "RC_RUNTIME_EVENT_ACHIEVEMENT_TRIGGERED",
+		"RC_RUNTIME_EVENT_ACHIEVEMENT_PRIMED", "RC_RUNTIME_EVENT_LBOARD_STARTED",
+		"RC_RUNTIME_EVENT_LBOARD_CANCELED", "RC_RUNTIME_EVENT_LBOARD_UPDATED",
+		"RC_RUNTIME_EVENT_LBOARD_TRIGGERED", "RC_RUNTIME_EVENT_ACHIEVEMENT_DISABLED",
+		"RC_RUNTIME_EVENT_LBOARD_DISABLED"};
+	const char* event_text =
+		((unsigned)runtime_event->type >= std::size(events)) ? "unknown" : events[(unsigned)runtime_event->type];
+	DevCon.WriteLn("Cheevos Event %s for %u", event_text, runtime_event->id);
+
+	if (runtime_event->type == RC_RUNTIME_EVENT_ACHIEVEMENT_TRIGGERED)
+		UnlockAchievement(runtime_event->id);
+	else if (runtime_event->type == RC_RUNTIME_EVENT_LBOARD_TRIGGERED)
+		SubmitLeaderboard(runtime_event->id, runtime_event->value);
+}
+
+unsigned Achievements::CheevosPeekEE(unsigned address, unsigned num_bytes, void* ud)
+{
+	switch (num_bytes)
+	{
+		case 1:
+			return static_cast<unsigned>(static_cast<u32>(vtlb_ramRead<u8>(address)));
+
+		case 2:
+			return static_cast<unsigned>(static_cast<u32>(vtlb_ramRead<u16>(address)));
+
+		case 4:
+			return static_cast<unsigned>(vtlb_ramRead<u32>(address));
+
+		default:
+			return 0;
+	}
+}
+
+#ifdef ENABLE_RAINTEGRATION
+
+#include "RA_Consoles.h"
+
+namespace Achievements::RAIntegration
+{
+	static void InitializeRAIntegration(void* main_window_handle);
+
+	static int RACallbackIsActive();
+	static void RACallbackCauseUnpause();
+	static void RACallbackCausePause();
+	static void RACallbackRebuildMenu();
+	static void RACallbackEstimateTitle(char* buf);
+	static void RACallbackResetEmulator();
+	static void RACallbackLoadROM(const char* unused);
+	static unsigned char RACallbackReadEEMemory(unsigned int address);
+	static unsigned char RACallbackReadIOPMemory(unsigned int address);
+	static void RACallbackWriteEEMemory(unsigned int address, unsigned char value);
+	static void RACallbackWriteIOPMemory(unsigned int address, unsigned char value);
+} // namespace Achievements::RAIntegration
+
+void Achievements::SwitchToRAIntegration()
+{
+	g_using_raintegration = true;
+	g_raintegration_initialized = false;
+	g_active = true;
+	g_logged_in = true;
+}
+
+void Achievements::RAIntegration::InitializeRAIntegration(void* main_window_handle)
+{
+	RA_InitClient((HWND)main_window_handle, "PCSX2", GIT_TAG);
+	RA_SetUserAgentDetail(Achievements::GetUserAgent().c_str());
+
+	RA_InstallSharedFunctions(RACallbackIsActive, RACallbackCauseUnpause, RACallbackCausePause, RACallbackRebuildMenu,
+		RACallbackEstimateTitle, RACallbackResetEmulator, RACallbackLoadROM);
+	RA_SetConsoleID(PlayStation2);
+
+	// For now, only the EE memory is exposed.. and only the bottom 2GB of it :S
+	RA_InstallMemoryBank(0, RACallbackReadEEMemory, RACallbackWriteEEMemory, std::numeric_limits<int>::max());
```

Comment:
You should only expose the 32MB of system memory. We don't need the entire addressable space.

Reply:
You do, because the addressing is virtual; games can remap the addresses how they like. Plus you have the IOP memory and scratchpad ;) This actually creates an issue since the size is signed, only the lower 2GB is accessible.

Reply:
Every other system that RetroAchievements supports that has more than only 64KB of addressable memory only accesses the RAM chips (and possibly the SRAM chips) - not the full address space.

I understand that the code can access the RAM from various addresses to provide slightly different behavior, but the system still only has 32MB of RAM, so when we're looking for how many lives the player has left, we don't need to look at the ROM, echoes caused by disconnected pins, the video RAM, the memory cards, or whatever else may be virtualized through the bus. We only need the volatile memory that will be mutated by the program.

Because of the way we expect memory to be exposed, we don't expect our addresses to line up 1:1 with the addresses as they would appear on the system. For static memory, we just reference the value at the location where we expect to find it. For traversing a pointer, we manually calculate the adjustment needed to turn the real pointer into a bank address (usually by just masking off the uppermost bits).

Also, the search code will not handle this well. To find changes in memory, we take a snapshot of the exposed RAM and compare the current values to that snapshot. You've just provided us with a 2GB snapshot.

Reply:
I still suspect you're going to run into issues dereferencing pointers that are in the high address space, and you're assuming that the game will set up a 1:1 mapping between physical and virtual addresses (remember, the PS2 has a TLB and address translation).

The reason I went with virtual addresses is because that's what other external tools (through PINE) use, and it's the only way you can access IOP memory and the scratchpad.

But I'll change it for now, and if it creates issues we can deal with it later on I guess.

Reply:
I will just add that trying to find stuff with physical addresses on the PS2 could be a real ball ache depending on what ytou wanna do, the chance of there being 1:1 mapping is highly unlikely, unless using your tools to make the achievements bypass any needing to debug it yourself to work out what's going on, it's going to make the life of people making the achievements harder.

Reply:
FWIW the physical address route will probably complicate relying on addresses obtained not from a memory scanner, but from a disassembler. In an ideal world, retroachievements could expose an inbuilt function to convert a virtual address to a physical address and vice versa - on PS1 you could e.g. get away by just doing `+/- 0x80000000` to the address, but by the sounds of it PS2 isn't that simple.

Reply:
Having a callback to translate a virtual to physical address is definitely something we could provide on our side.

Reply:
And vice versa, yeah. Other platforms could benefit from it too, e.g. allowing to drop that "mask highest bits" hack on the PS1.

Reply:
Going physical to virtual is a bit trickier, as we don't keep track of reverse mappings (after all, the things which work with physical addresses, such a DMA, don't need virtual addresses in the first place). Plus, the fact that you can have many virtual mappings for one physical address (for fastmem, I maintained a list).

Reply:
True, and since racheevos don't ever **write** you are unlikely to ever need to go this route. But for **reading**, virtual-to-physical is very much needed.